{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import subprocess\n",
    "\n",
    "index = ['0','1']\n",
    "options = ['yes', 'no']\n",
    "\n",
    "\n",
    "tools_rna = pd.DataFrame(data=np.array([['HISAT2', 'StringTie', 'ballgown']]), columns=['Aligner', 'Expression Modeller', 'Differential Expression Tools'],index = index)\n",
    "def_params = pd.DataFrame(data=np.array([['-p 1', '--dta', '--rna-strandness RF'], ['-p 1', '--merge', '--rf', '-e', '-B'], []]), columns=['Aligner', 'Expression Modeller', 'Differential Expression Tools'])\n",
    "\n",
    "\n",
    "print(\"\\n Welcome to Icomic pipeline!\\n \\n Are you working with DNA-Seq or RNA-Seq data?? \\n \")\n",
    "\n",
    "d = {'Data': ['DNA-Seq','RNA-Seq']}\n",
    "r = pd.DataFrame(data = d)\n",
    "print(r)\n",
    "pipe= int(input(\"\\n Choose the numbers corresponding to your input \\n\"))\n",
    "for column in r:\n",
    "    print(\"\\n\\n You are working with \" + r.iloc[pipe][column] + \" data\\n\")\n",
    "    if r.iloc[pipe][column] == 'DNA-Seq':\n",
    "        tools = pd.DataFrame(data=np.array([['BWA_MEM','GATK_HC' , 'SnpEff'], ['Bowtie2', 'bcftools_call' , 'Annovar']]), columns=['Aligner', 'Variant_caller', 'Annotator'],index = index)\n",
    "\n",
    "        def_params = pd.DataFrame(data=np.array([[['-k 19' ,'-t 1', '-w 100', '-d 100'], ['-mbq 10', '-minReadsPerAlignStart 10', '-ploidy 2'], ['-i VCF', '-o VCF', '-t 1', 'variants_file STDIN]']], [['-p 1', '--ma 2', '--mp 6', '--np 1', '--rdg 5', '--rfg 5'], ['--threads 1', '-O v', '-s all samples'], ['-thread 1']]]), columns=['Aligner', 'Variant_caller', 'Annotator'])\n",
    "\n",
    "        gatk = np.array(['HaplotypeCaller', 'GenotypeGVCFs', 'BaseRecalibrator', 'VariantRecalibrator'])\n",
    "        \n",
    "\n",
    "\n",
    "        files = ['samples', 'units']\n",
    "        ref_files = ['genome ', 'known-variants']\n",
    "\n",
    "        conf = open('config.yaml',\"w\") \n",
    "        for file in files:\n",
    "            path = input(\"\\n \\n Enter the {} file name with path, (eg: /path/to/filename.extension) : \\t\".format(file))\n",
    "            conf.write(file + \": \" + path + \"\\n\")\n",
    "        conf.write(\"ref: \" + \"\\n\")\n",
    "        name = input(\"\\n \\n Enter the reference name (for SnpEff) : \\t\")\n",
    "        conf.write(\"  name: \" + name + \"\\n\")\n",
    "        for ref_file in ref_files:\n",
    "            ref_path = input(\"\\n \\n Enter the reference {} file name with path, (eg: /path/to/filename.extension) : \\t\".format(ref_file))\n",
    "            conf.write(\"  \" + ref_file + \": \" + ref_path + \"\\n\")\n",
    "        conf.write(\"processing: \\n\")\n",
    "        conf.write(\"  \"+ \"remove-duplicates: true\\n\")\n",
    "        #cap_reg = input(\"\\n\\n Enter the restriction-regions file name with path (eg: /path/to/filename.extension) : \\t\")\n",
    "        #conf.write(\"  \"+\"restrict-regions: \" + cap_reg + \"\\n\" )\n",
    "        conf.write(\"filtering:\\n\")\n",
    "        conf.write(\"  \"+\"vqsr: false\\n\")\n",
    "        conf.write(\"  \"+\"hard:\\n\")\n",
    "        conf.write(\"    \"+\"snvs:\\n\")\n",
    "        conf.write(\"      \" + '\"QD < 2.0 || FS > 60.0 || MQ < 40.0 || MQRankSum < -12.5 || ReadPosRankSum < -8.0\"\\n')\n",
    "        conf.write(\"    \"+\"indels:\\n\")\n",
    "        conf.write(\"      \"+'\"QD < 2.0 || FS > 200.0 || ReadPosRankSum < -20.0\"\\n')\n",
    "        conf.write(\"params: \" + \"\\n\")\n",
    "        for col in tools:\n",
    "            for number in [0,1]:\n",
    "                if tools.iloc[number][col] != 'GATK_HC':\n",
    "                    conf.write(\"  \" + tools.iloc[number][col] + ': \"\" \\n')\n",
    "                elif tools.iloc[number][col] == \"GATK_HC\":\n",
    "                    conf.write('  GATK_HC : \\n')\n",
    "                    for i in range(len(gatk)):\n",
    "                        conf.write('    ' + gatk[i] + ': \"\"\\n')\n",
    "        conf.write('  cutadapt: \"\" \\n')\n",
    "        conf.write(\"  picard:\\n\")\n",
    "        conf.write('    MarkDuplicates: \"REMOVE_DUPLICATES=true\"\\n')\n",
    "        conf.close() \n",
    "\n",
    "        snake = open ('Snakefile',\"w\")\n",
    "        snake.write('include: \"rules/common.smk\"\\n\\n')\n",
    "        snake.write('rule all:\\n')\n",
    "        snake.write('  input:\\n')\n",
    "        snake.write('    \"qc/multiqc.html\"\\n')\n",
    "        snake.write('include: \"rules/cutadapt.smk\"\\n')\n",
    "        snake.write('include: \"rules/qc.smk\"\\n')\n",
    "        snake.close()\n",
    "\n",
    "\n",
    "        print(tools)\n",
    "\n",
    "while True:\n",
    "    answer_qc = input(\"\\n \\n Have you done quality check? \\t\") or \"yes\"\n",
    "    if (answer_qc not in options):\n",
    "        print(\"\\n Your entry should be either 'yes' or 'no'\")\n",
    "        continue\n",
    "    if answer_qc == \"no\":\n",
    "        print(\"\\n \\n Running FastQC!! Check qc results in the folder qc/fastqc !!\")\n",
    "        subprocess.run([\"snakemake\", \"--use-conda\", \"-U\", \"fastqc\"])\n",
    "    print(\"\\n \\n Preparing to run cutadapt with default parameters ['-j 1', '-e 0.1', '-n 1', '-O 3']!! Check the qc results after trimming in the folder qc/fastqc !! \")\n",
    "    while True:\n",
    "        answer = input(\"\\n \\n Do you want to run cutadapt with default parameters? \\t\") or \"yes\"\n",
    "        if (answer not in options):\n",
    "            print(\"\\n Your entry should be either 'yes' or 'no'\")\n",
    "            continue\n",
    "        if answer == \"no\":\n",
    "            cut_params_list = input(\"\\n\\n Press enter to list out the parameters for cutadapt : \")\n",
    "            if cut_params_list == \"\":\n",
    "                a = open(\"cutadapt.txt\", \"r\")\n",
    "                print(a.read())\n",
    "#            else:\n",
    "#                print(\"Press the enter key!!!\")\n",
    "#                continue\n",
    "            try:\n",
    "                o = int(input(\"\\n How many parameters do you want to add for cutadapt : \\t\"))\n",
    "            except ValueError:\n",
    "                print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                continue\n",
    "\n",
    "            cut_new_params = []\n",
    "            for j in range(1,o+1):\n",
    "                cut_add_params = (input(\"\\n Enter the additional parameter for cutadapt, for example j 4 : \\t\"))\n",
    "                cut_new_params.append(cut_add_params)\n",
    "\n",
    "            r = open('config.yaml').read()\n",
    "            r = r.replace('  cutadapt: \"\" \\n', \"  cutadapt: \" + ' '.join(map(lambda x: '\"-' + x + '\"\\n', cut_new_params)) + \"\\n\")\n",
    "            y = open(\"config.yaml\", 'w')\n",
    "            y.write(r)\n",
    "            y.close()\n",
    "        elif (answer == \"yes\"):\n",
    "            print(\"\\n \\n You have chosen to work with the parameters ['-j 1', '-e 0.1', '-n 1', '-O 3'] !!\")\n",
    "            subprocess.run([\"snakemake\", \"--use-conda\", \"-U\", \"fastqc_after\"])\n",
    "        answer_align = input(\"\\n \\n Do you want to Proceed with alignment? \\t\") or \"yes\"\n",
    "        if answer_align == \"no\":\n",
    "            continue\n",
    "        else:\n",
    "            if r.iloc[pipe][column] == 'DNA-Seq':\n",
    "                for col in tools.columns:\n",
    "                    while True:\n",
    "                        try:\n",
    "                            number = int( input(\" \\n\\n Enter the number associated with your choice of {} : \\t\".format(col)))\n",
    "                        except ValueError:\n",
    "                            print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                            continue\n",
    "\n",
    "                        if number >= len(tools):\n",
    "                            print(\"\\n Invalid choice of\", col ,\"!!!\")\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            print(\"\\n You have chosen the \", col ,\",\" ,tools.iloc[number][col])\n",
    "                            with open('Snakefile','a') as snakefile:\n",
    "                                snakefile.write('include: \"rules/' + tools.iloc[number][col] + '.smk\" \\n')\n",
    "                                snakefile.close()\n",
    "                            print(\"\\n The default parameters for \", col, tools.iloc[number][col] , \" are \" , def_params.iloc[number][col])\n",
    "                            params_list = input(\"\\n \\n Press enter to list out the parameters : \")\n",
    "                            if params_list == \"\":\n",
    "                                    f = open(tools.iloc[number][col]+\".txt\", \"r\")\n",
    "                                    print(f.read())\n",
    "    #                        else:\n",
    "    #                            print(\"\\n \\n Press the enter key alone!!!\")\n",
    "    #                            continue\n",
    "\n",
    "                            while True:\n",
    "                                answer = input(\"\\n \\n Do you want to work with default parameters? \\t\") or \"yes\"\n",
    "                                if (answer not in options):\n",
    "                                    print(\"\\n Your entry should be either 'yes' or 'no'\")\n",
    "                                    continue\n",
    "                                if answer == \"no\":\n",
    "                                    if tools.iloc[number][col] == 'GATK_HC':\n",
    "                                        for i in range(len(gatk)):\n",
    "                                            gatk_params_list = input(\"\\n\\n Press enter to list out the parameters for \" + gatk[i] + \": \")\n",
    "                                            if gatk_params_list == \"\":\n",
    "                                                f = open(gatk[i] +\".txt\", \"r\")\n",
    "                                                print(f.read())\n",
    "    #                                        else:\n",
    "    #                                            print(\"Press the enter key!!!\")\n",
    "                #                                continue\n",
    "                                                try:\n",
    "                                                    g = int(input(\"\\n How many parameters do you want to add for {} : \\t\".format(gatk[i])))\n",
    "                                                except ValueError:\n",
    "                                                    print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                                                    continue\n",
    "\n",
    "                                                gatk_new_params = []\n",
    "                                                for y in range(1,g+1):\n",
    "                                                    gatk_add_params = (input(\"\\n Enter the additional parameter for {}, for example j 4 : \\t\".format(gatk[i])))\n",
    "                                                    gatk_new_params.append(gatk_add_params)\n",
    "                                                print(gatk_new_params)\n",
    "                                                k = open('config.yaml').read()\n",
    "                                                k = k.replace('    ' + gatk[i] + ': \"\"\\n', \"    \" + gatk[i] + \": \" + ' '.join(map(lambda x: '\"-' + x + '\"\\n', gatk_new_params)) + \"\\n\")\n",
    "                                                t = open(\"config.yaml\", 'w')\n",
    "                                                t.write(k)\n",
    "                                                t.close()\n",
    "\n",
    "                                    else:\n",
    "                                        try:\n",
    "                                            p = int(input(\"\\n How many parameters do you want to add? \\t\"))\n",
    "                                        except ValueError:\n",
    "                                            print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                                            continue\n",
    "                                        new_params = []\n",
    "                                        for r in range(1,p+1):\n",
    "                                            add_params = (input(\"\\n Enter the additional parameter {for example j 4 } : \\t\"))\n",
    "                                            new_params.append(add_params)\n",
    "\n",
    "\n",
    "                                    w = open('config.yaml').read()\n",
    "                                    w = w.replace(\"\\t\" + tools.iloc[number][col] + ': \"\" \\n', \"\\t\" + tools.iloc[number][col] + \": \" + ' '.join(map(lambda x: '\"-' + x + '\"\\n', new_params)) + \"\\n\")\n",
    "                                    fl = open(\"config.yaml\", 'w')\n",
    "                                    fl.write(w)\n",
    "                                    fl.close()\n",
    "                                    all_params = def_params.iloc[number][col] + (new_params)\n",
    "\n",
    "\n",
    "                                    print(\"\\n \\n You have chosen to work with \", col , tools.iloc[number][col] , \" with the parameters \" , all_params)\n",
    "                                    break\n",
    "                                elif (answer == \"yes\"):\n",
    "                                    print(\"\\n \\n You have chosen to work with \", col , tools.iloc[number][col], \" with the parameters \" , def_params.iloc[number][col])\n",
    "\n",
    "                                    break\n",
    "                            break  \n",
    "            elif r.iloc[pipe][column] == 'RNA-Seq':\n",
    "                for col in tools_rna.columns:\n",
    "                    while True:\n",
    "                        try:\n",
    "                            number = int( input(\" \\n\\n Enter the number associated with your choice of {} : \\t\".format(col)))\n",
    "                        except ValueError:\n",
    "                            print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                            continue\n",
    "\n",
    "                        if number >= len(tools):\n",
    "                            print(\"\\n Invalid choice of\", col ,\"!!!\")\n",
    "                            continue\n",
    "\n",
    "                        else:\n",
    "                            print(\"\\n You have chosen the \", col ,\",\" ,tools_rna.iloc[number][col])\n",
    "                            with open('Snakefile','a') as snakefile:\n",
    "                                snakefile.write('include: \"rules/' + tools.iloc[number][col] + '.smk\" \\n')\n",
    "                                snakefile.close()\n",
    "                            print(\"\\n The default parameters for \", col, tools_rna.iloc[number][col] , \" are \" , def_params.iloc[number][col])\n",
    "                            params_list = input(\"\\n \\n Press enter to list out the parameters : \")\n",
    "                            if params_list == \"\":\n",
    "                                    f = open(tools_rna.iloc[number][col]+\".txt\", \"r\")\n",
    "                                    print(f.read())\n",
    "    #                        else:\n",
    "    #                            print(\"\\n \\n Press the enter key alone!!!\")\n",
    "    #                            continue\n",
    "\n",
    "                            while True:\n",
    "                                answer = input(\"\\n \\n Do you want to work with default parameters? \\t\") or \"yes\"\n",
    "                                if (answer not in options):\n",
    "                                    print(\"\\n Your entry should be either 'yes' or 'no'\")\n",
    "                                    continue\n",
    "                                if answer == \"no\":\n",
    "                                    try:\n",
    "                                        p = int(input(\"\\n How many parameters do you want to add? \\t\"))\n",
    "                                    except ValueError:\n",
    "                                        print(\"\\n Sorry, entry should be a positive integer!!\")\n",
    "                                        continue\n",
    "                                    new_params = []\n",
    "                                    for r in range(1,p+1):\n",
    "                                        add_params = (input(\"\\n Enter the additional parameter {for example j 4 } : \\t\"))\n",
    "                                        new_params.append(add_params)\n",
    "\n",
    "\n",
    "                                    w = open('config.yaml').read()\n",
    "                                    w = w.replace(\"\\t\" + tools.iloc[number][col] + ': \"\" \\n', \"\\t\" + tools.iloc[number][col] + \": \" + ' '.join(map(lambda x: '\"-' + x + '\"\\n', new_params)) + \"\\n\")\n",
    "                                    fl = open(\"config.yaml\", 'w')\n",
    "                                    fl.write(w)\n",
    "                                    fl.close()\n",
    "                                    all_params = def_params.iloc[number][col] + (new_params)\n",
    "\n",
    "                                    print(\"\\n \\n You have chosen to work with \", col , tools.iloc[number][col] , \" with the parameters \" , all_params)\n",
    "                                    break\n",
    "                                elif answer == \"yes\":\n",
    "                                    print(\"\\n \\n You have chosen to work with \", col , tools.iloc[number][col], \" with the parameters \" , def_params.iloc[number][col])\n",
    "\n",
    "                                    break\n",
    "                        break  \n",
    "                    \n",
    "                \n",
    "                \n",
    "                break\n",
    "\n",
    "        \n",
    "        \n",
    "        break\n",
    "\n",
    "    #        break \n",
    "if r.iloc[pipe][column] == 'DNA-Seq':\n",
    "    with open('Snakefile', 'r+') as fd:\n",
    "        contents = fd.readlines()\n",
    "        if tools.iloc[number]['Annotator'] == 'Annovar':\n",
    "            contents.insert(4, '    \"annotated/all.vcf.variant_function\", \\n')  # new_string should end in a newline\n",
    "            fd.seek(0)  # readlines consumes the iterator, so we need to start over\n",
    "            fd.writelines(contents) \n",
    "        else:\n",
    "            contents.insert(4, '    \"annotated/all.vcf.gz\",\\n')\n",
    "            fd.seek(0)  # readlines consumes the iterator, so we need to start over\n",
    "            fd.writelines(contents)\n",
    "    subprocess.run([\"snakemake\", \"--use-conda\"])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
